src/main/attached_database.cpp

```diff
@@ -0,0 +1,35 @@
+#include "main/attached_database.h"
+
+#include "main/db_config.h"
+#include "storage/storage_manager.h"
+#include "transaction/transaction_manager.h"
+
+namespace kuzu {
+namespace main {
+
+void AttachedDatabase::invalidateCache() {
+    if (dbType != common::ATTACHED_KUZU_DB_TYPE) {
+        auto catalogExtension = catalog->ptrCast<extension::CatalogExtension>();
+        catalogExtension->invalidateCache();
+    }
+}
+
+AttachedKuzuDatabase::AttachedKuzuDatabase(std::string dbPath, std::string dbName,
+    std::string dbType, ClientContext* clientContext)
+    : AttachedDatabase{std::move(dbName), std::move(dbType), nullptr /* catalog */} {
+    auto vfs = clientContext->getVFSUnsafe();
+    auto path = vfs->expandPath(clientContext, dbPath);
+    if (!vfs->fileOrPathExists(path)) {
+        throw common::RuntimeException(common::stringFormat(
+            "Cannot attach a remote kuzu database due to invalid path: {}.", dbPath));
+    }
```

Comment:
Can u also check if the wal file is empty? We should not replay wal to recover on read only database.